openformation/strawman
# feat(strawman-cli): Implements `rc` command

Closes: #9 

## How to test

1. Since we do not have an install script yet (see: #22), it's best to set up an alias for strawman first:
```bash
alias strawman="deno run --unstable --allow-read --allow-write --allow-net modules/strawman-cli/cli.ts"
```

2. Start a strawman server in replay mode:
```bash
strawman start -e -m replay https://httpbin.org
```

### `strawman rc logs`

Run
```
strawman rc logs
```

You should see the output:
```
███████╗████████╗██████╗  █████╗ ██╗    ██╗███╗   ███╗ █████╗ ███╗   ██╗
██╔════╝╚══██╔══╝██╔══██╗██╔══██╗██║    ██║████╗ ████║██╔══██╗████╗  ██║
███████╗   ██║   ██████╔╝███████║██║ █╗ ██║██╔████╔██║███████║██╔██╗ ██║
╚════██║   ██║   ██╔══██╗██╔══██║██║███╗██║██║╚██╔╝██║██╔══██║██║╚██╗██║
███████║   ██║   ██║  ██║██║  ██║╚███╔███╔╝██║ ╚═╝ ██║██║  ██║██║ ╚████║
╚══════╝   ╚═╝   ╚═╝  ╚═╝╚═╝  ╚═╝ ╚══╝╚══╝ ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝  ╚═══╝

Strawman Server 1.0.0-alpha.1
```

The process should not have exited.

The terminal that runs the strawman server should have logged:
```
[⚐] A remote control client has been connected to Strawman.
```

Open another terminal and send a request to the strawman server:
```
curl http://localhost:8080/
```

The output of the terminal that runs `strawman rc logs` should have changed to:
```
███████╗████████╗██████╗  █████╗ ██╗    ██╗███╗   ███╗ █████╗ ███╗   ██╗
██╔════╝╚══██╔══╝██╔══██╗██╔══██╗██║    ██║████╗ ████║██╔══██╗████╗  ██║
███████╗   ██║   ██████╔╝███████║██║ █╗ ██║██╔████╔██║███████║██╔██╗ ██║
╚════██║   ██║   ██╔══██╗██╔══██║██║███╗██║██║╚██╔╝██║██╔══██║██║╚██╗██║
███████║   ██║   ██║  ██║██║  ██║╚███╔███╔╝██║ ╚═╝ ██║██║  ██║██║ ╚████║
╚══════╝   ╚═╝   ╚═╝  ╚═╝╚═╝  ╚═╝ ╚══╝╚══╝ ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝  ╚═══╝

Strawman Server 1.0.0-alpha.1

[➘] [GET] /
  → Received request
[☓] [GET] /
  → #1641473907: Template was not found.
         at Object.replayRequest (file:///.../modules/strawman-core/application/service/replayRequest.ts:57:21)
         at Object.handleRequest (file:///.../modules/strawman-core/application/strawman.ts:79:22)
         at handleProxyRequest (file:///.../strawman/modules/strawman-cli/lib/http/handleProxyRequest.ts:36:44)
         at serveHttp (file:///.../modules/strawman-cli/lib/http/serveHttp.ts:44:9)
```

### `strawman rc capture`

Run
```
strawman rc capture
```

You should see the output:
```
███████╗████████╗██████╗  █████╗ ██╗    ██╗███╗   ███╗ █████╗ ███╗   ██╗
██╔════╝╚══██╔══╝██╔══██╗██╔══██╗██║    ██║████╗ ████║██╔══██╗████╗  ██║
███████╗   ██║   ██████╔╝███████║██║ █╗ ██║██╔████╔██║███████║██╔██╗ ██║
╚════██║   ██║   ██╔══██╗██╔══██║██║███╗██║██║╚██╔╝██║██╔══██║██║╚██╗██║
███████║   ██║   ██║  ██║██║  ██║╚███╔███╔╝██║ ╚═╝ ██║██║  ██║██║ ╚████║
╚══════╝   ╚═╝   ╚═╝  ╚═╝╚═╝  ╚═╝ ╚══╝╚══╝ ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝  ╚═══╝

Strawman Server 1.0.0-alpha.1

Strawman mode was set to "capture".
```

The terminal that runs the strawman server should have logged:
```
[⚐] A remote control client has been connected to Strawman.
[⚐] Strawman is in capture mode.
[⚐] A remote control client has been disconnected from Strawman.
```

Strawman should actually be in capture mode now. Try capturing a response and see, if a template is being written to `snapshots/`.

Running `strawman rc replay` puts strawman back into "replay" mode.

### `strawman rc delete`

Make sure strawman is in capture mode by running `strawman rc capture` again.


Open another terminal and send a request to the strawman server:
```
curl http://localhost:8080/
```

This should have created the file `./snapshots/GET.mock.ts`.

Run
```
strawman rc delete GET /
```

The output should be:
```
███████╗████████╗██████╗  █████╗ ██╗    ██╗███╗   ███╗ █████╗ ███╗   ██╗
██╔════╝╚══██╔══╝██╔══██╗██╔══██╗██║    ██║████╗ ████║██╔══██╗████╗  ██║
███████╗   ██║   ██████╔╝███████║██║ █╗ ██║██╔████╔██║███████║██╔██╗ ██║
╚════██║   ██║   ██╔══██╗██╔══██║██║███╗██║██║╚██╔╝██║██╔══██║██║╚██╗██║
███████║   ██║   ██║  ██║██║  ██║╚███╔███╔╝██║ ╚═╝ ██║██║  ██║██║ ╚████║
╚══════╝   ╚═╝   ╚═╝  ╚═╝╚═╝  ╚═╝ ╚══╝╚══╝ ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝  ╚═══╝

Strawman Server 1.0.0-alpha.1

Template GET / was deleted.
```

And `./snapshots/GET.mock.ts` should be gone.
